server/game/gameSystems/InitiateAttackWithUnitSystem.ts

```diff
@@ -0,0 +1,76 @@
+import type { Card } from '../core/card/Card';
+import AbilityResolver from '../core/gameSteps/AbilityResolver';
+import type { TriggeredAbilityContext } from '../core/ability/TriggeredAbilityContext';
+import { CardTargetSystem, ICardTargetSystemProperties } from '../core/gameSystem/CardTargetSystem';
+import { UnitCard } from '../core/card/CardTypes';
+import { InitiateAttackAction } from '../actions/InitiateAttackAction';
+import { AbilityContext } from '../core/ability/AbilityContext';
+import Contract from '../core/utils/Contract';
+import { IAttackProperties } from './AttackStepsSystem';
+import * as GameSystemLibrary from './GameSystemLibrary';
+
+export interface IInitiateUnitAttackProperties extends IAttackProperties {
```

Comment:
The "unit" and "WithUnit" parts of this seem redundant. Or is this to signify that it will be attacking with a specifically chosen unit as opposed to just declaring an attack?

Reply:
Yeah I really wrestled with the naming here. Basically there's two closely related but distinctly different systems:

- The one that walks through the attack steps and manages all of the stages from beginning to end. This accepts the attacker as a property (so it has to be already decided) and the target is the attack target, so a target resolver would be presented to decide the attack target if needed. This is currently called `AttackStepsSystem` iirc.
- An outer system (this one) which allows you to choose both the attacker and the defender in two different steps. It's functionally allowing you to choose a unit and then trigger an attack with that unit, which ends up calling the `AttackStepsSystem` above. I am currently calling it `InitiateAttackWithUnit` to indicate that you are choosing the attacking unit, but I don't love that name and am very open to suggestions

Reply:
What about something like TriggeredAttackSystem for this one?

Reply:
I prefer using `Initiate` since it's calling the `InitiateAttackAbility` class (which is what's called when a user clicks on the card and attacks). Or else change both to `Trigger`.

Bigger picture though, I'm not sure if that clear disambiguates from the other attack system since they're both triggering / initiating an attack

Reply:
Eh, actually maybe it's clear enough, since `AttackSteps` is not really ever going to be used anyway except in some very special cases